src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1703,47 +1703,75 @@ abstract class RefChecks extends Transform {
 
           transform(qual)
       case Apply(fn, args) =>
+        currentApplication = tree
         // sensicality should be subsumed by the unreachability/exhaustivity/irrefutability
         // analyses in the pattern matcher
         if (!inPattern) {
           checkImplicitViewOptionApply(tree.pos, fn, args)
           checkSensible(tree.pos, fn, args) // TODO: this should move to preEraseApply, as reasoning about runtime semantics makes more sense in the JVM type system
           checkNamedBooleanArgs(fn, args)
         }
-        currentApplication = tree
         tree
     }
 
     /** Check that boolean literals are passed as named args.
-     *  The rule is enforced when the type of the parameter is `Boolean`.
-     *  The rule is relaxed when the method has exactly one boolean parameter
+     *  The rule is enforced when the type of the parameter is `Boolean`,
+     *  and there is more than one parameter with an unnamed argument.
+     *  The stricter internal lint warns for any unnamed argument,
+     *  except that the rule is relaxed when the method has exactly one boolean parameter
      *  and it is the first parameter, such as `assert(false, msg)`.
      */
     private def checkNamedBooleanArgs(fn: Tree, args: List[Tree]): Unit = {
       val sym = fn.symbol
       def applyDepth: Int = {
         def loop(t: Tree, d: Int): Int =
           t match {
-            case Apply(f, _) => loop(f, d+1)
+            case Apply(t, _) => loop(t, d+1)
             case _ => d
           }
         loop(fn, 0)
       }
-      def isAssertParadigm(params: List[Symbol]): Boolean = !sym.isConstructor && !sym.isCaseApplyOrUnapply && {
-        params match {
-          case h :: t => h.tpe == BooleanTpe && !t.exists(_.tpe == BooleanTpe)
-          case _ => false
-        }
-      }
-      if (settings.lintNamedBooleans && !sym.isJavaDefined && !args.isEmpty) {
+      if (settings.warnNamedLiteral.value && !sym.isJavaDefined && !args.isEmpty) {
+        def isUnnamedArg(t: Tree) = t.hasAttachment[UnnamedArg.type]
+        def isDefaultArg(t: Tree) = t.symbol != null && (
+          t.symbol.isDefaultGetter || t.symbol.name.startsWith(nme.NAMEDARG_PREFIX) && {
+            analyzer.NamedApplyBlock.namedApplyInfo(currentApplication) match {
+              case Some(analyzer.NamedApplyInfo(_, _, _, _, original)) =>
+                val treeInfo.Applied(_, _, argss) = original
+                val orig = argss.head(t.symbol.name.decoded.stripPrefix(nme.NAMEDARG_PREFIX).toInt - 1)
```

Comment:
this is totally wrong. `x$42` is just a fresh name, not an index into the param list.